src/impl/metadata/retriever/baseMetadataRetriever.ts

```diff
@@ -1,111 +1,255 @@
-        `Method: isToolingandLimitBasedQueryRetrieval : true, QUERY:  ${this.query}`,
-        LoggerLevel.TRACE
-      );
-
-      let retrievedRecordSize = 0;
-      let offset = 0;
-      this.totalSize = await this.getCount();
-
-      while (retrievedRecordSize < this.totalSize) {
-        SFPowerkit.log(
-          `To Retrieve Total Size:  ${this.totalSize},Retrieved Size:   ${retrievedRecordSize} , Current Offset: ${offset}`,
-          LoggerLevel.TRACE
-        );
-
-        let result: QueryResult<T>;
-        SFPowerkit.log(this.queryWithOffsetsAndLimit, LoggerLevel.TRACE);
-        result = await conn.tooling.query<T>(this.queryWithOffsetsAndLimit);
-        retrievedRecordSize += result.totalSize;
-
-        records.push(...result.records);
+  private generateCountQuery() {
+    let queryParts = this.query.toUpperCase().split("FROM");
+    let objectParts = queryParts[1].trim().split(" ");
+    let objectName = objectParts[0].trim();
+    this.objectName = objectName;
 
-        offset = offset + this.fetchSize;
-        this.queryWithOffsetsAndLimit = this.query.concat(
-          ` LIMIT ${this.fetchSize} OFFSET ${offset}`
-        );
-      }
-    }
+    let countQuery = `SELECT COUNT() FROM ${objectName}`;
 
-    return records;
+    return countQuery;
   }
 
-  private async queryMore(url: string): Promise<QueryResult<T>> {
+  protected async getObjects(): Promise<T[]> {
+    //let records: T[] = [];
     const conn = this.org.getConnection();
-    let result: QueryResult<T>;
+
     if (this.tooling) {
-      result = await conn.tooling.queryMore<T>(url);
+      return executeToolingQueryAsync(this.query, conn, this.objectName);
     } else {
-      result = await conn.queryMore<T>(url);
+      let recordsCount = await this.getCount();
+      if (recordsCount > BULK_THRESHOLD) {
+        return executeBulkQueryAsync(
+          this.query,
+          conn,
+          this.objectName,
+          recordsCount
+        );
+      } else {
+        return executeQueryAsync(this.query, conn, this.objectName);
+      }
     }
-    return result;
   }
 
   private async getCount() {
     SFPowerkit.log(`Count Query: ${this.countQuery}`, LoggerLevel.TRACE);
-    let result = await this.org.getConnection().tooling.query(this.countQuery);
+    let result = await this.org.getConnection().query(this.countQuery);
     SFPowerkit.log(`Retrieved count ${result.totalSize}`, LoggerLevel.TRACE);
     return result.totalSize;
   }
 }
+
+export async function executeToolingQueryAsync(
+  query,
+  conn,
+  object
+): Promise<any[]> {
+  let promiseQuery = new Promise<any[]>((resolve, reject) => {
+    let records = [];
+    let hasInitProgress = false;
+    let progressBar = undefined;
+    if (
+      SFPowerkit.logLevel === LoggerLevel.DEBUG ||
+      SFPowerkit.logLevel === LoggerLevel.TRACE
+    ) {
+      progressBar = cli.progress({
+        format: `Querying data from ${object} - PROGRESS  | {bar} | {value}/{total} Records fetched`,
```

Comment:
@genoud  Can we remove the cli depenedency from the impl layer?

Reply:
Yes we can do that. May be put it in the SFPowerkit class where we implemented the log.

Reply:
Sounds good, lets move it over there

Reply:
@genoud Thanks for the fix.